# <img border="0" src="https://www.svgrepo.com/show/1025/task.svg" width="40" height="40"> Projects

***

<br/>

Since each of the 4 group themes have completely different softwares, please follow in instructions specific to each one:

## RNA-seq hands-on (Day 1)
***

- [Instructions for running the GENE-SWitCH RNA-seq pipeline](http://genoweb.toulouse.inra.fr/~sdjebali/courses/SIB_august2020/instructions/1.pipeline/)
- [Instructions for running the RNA-seq downstream analyses](http://genoweb.toulouse.inra.fr/~sdjebali/courses/SIB_august2020/instructions/2.analyses/)

<br/>

<br/>

# Projects


## Single cell RNA analysis
***

Please follow the [pre-course instructions](precourse.md) in order to install all necessary software and packages.

The proposed educational plan for single cell RNA-seq data analysis project will be done using the Project-based learning (a.k.a. PBL) approach. This way, rather than having lectures on a topic and defined set of instructions and pre-given steps to follow (which focus on passive individual learning), you are required to learn the topic in a more **active** and **dynamic** way in order to solve the data analysis tasks as group. Therefore, each group is provided with:

1. The [report file](single_cell/code/project_report.md) containing a list of instructions and questions giving the overall direction and order of the steps needed to be done and;
2. A [glossary of terms](single_cell/glossary/glossary_of_terms_single_cell.md) which contains detailed information about many of the steps and provides further references on the topics.

In this manner, each group is required to gather information from the glossary to perform single cell data analysis by filling the report file (using this [project_report.Rmd](single_cell/code/project_report.Rmd) file), by:

1. Reading the instructions for a certain task;
2. Searching the glossary for an answer;
3. Discussing with your colleges which step needs to be done or included;
4. Replacing the instructions with a text explaining the rational for this step and;
5. Adding and running code from that step.

You are encouraged to use any other additional website / article in your report. That also includes addition of other previous experiences and/or additional code steps from other sources, as long as the rational for the inclusion is also discussed in the report text. You can also add any other plots and visualisations in your report to illustrate your results.

In the end, each group is expected to have 1 report file with all analysis steps and the rational for the following projects:

<br/>

#### PROJECT 1: Blocking myeloid development during colitis.

<details>
<summary>Click to expand!</summary>

  **(Background)** Ulcerative colitis (UC) is an inflammatory bowel disease (IBD) driven mainly by colonic innate inflammatory cells such as macrophages, monocytes and neutrophils (Czarnewski 2019, Skatteborg 2020). A recent study showed that patients that present higher neutrophilic/monocytic inflammatory signature (known as UC1) become refractory to both anti-TNF and anti-a4b7 integrin therapy (Czarnewski 2019, Skatteborg 2020), which leads to surgical intervention for removal of the colon. Both of these inflammatory cells are short lived and originate from the common myeloid progenitor (CMP) in the bone marrow and requires constant replenishment in order to sustain elevated cell number in the colon. Herein, our main goal is to identify potential gene candidates that can block either one or both pathways of neutrophil and monocyte differentiation in the bone marrow.

  **(Main research question)** Which genes specifically drive the differentiation of 1) common myeloid progenitors, 2) Monocytes and 3) Neutrophils.

  **(Importance)** Identifying such genes will allow us to: 1) perform experiment in Tamoxifen-transgenic mice where those cells can be depleted during the course of colitis. 2) find potential drugs that can inhibit those genes/pathways in order to block myeloid cell differentiation during colitis in mice (with priority to already approved drugs).

</details>

<br/>

#### PROJECT 2: Understanding Fibroblast function during colonic inflammation.

<details>
<summary>Click to expand!</summary>

  TO DO

</details>

<br/>

<br/>

## Long-read RNA sequencing
***

<br/>

<br/>

## Ribosome-profiling
***

<br/>

<br/>

## UV cross-linking immunoprecipitation (CLIP-seq)
***

<br/>

<br/>

### [Back to main](README.md)
